### PR TITLE
fix(api): Add @types/graphql dependency

### LINKED ISSUE #2362: Cannot use namespace 'DocumentNode' as a type.
I am getting below error when using this module. Does anyone know the root cause?

When running with webpack
ERROR in node_modules/@aws-amplify/api/lib/types/index.d.ts(45,21): error TS2709: Cannot use namespace 'DocumentNode' as a type.

AOT build error
ERROR in node_modules/@aws-amplify/api/lib/types/index.d.ts(45,21): error TS2709: Cannot use namespace 'DocumentNode' as a type.
node_modules/aws-amplify-angular/dist/src/components/xr/sumerian-scene-component/sumerian-scene.component.core.d.ts(14,16): error TS2709: Cannot use namespace 'AmplifyUI' as a type.
node_modules/aws-amplify-angular/dist/src/components/xr/sumerian-scene-component/sumerian-scene-loading.component.core.d.ts(6,16): error TS2709: Cannot use namespace 'AmplifyUI' as a type.
node_modules/aws-amplify-angular/dist/src/components/xr/sumerian-scene-component/sumerian-scene.component.ionic.d.ts(5,16): error TS2709: Cannot use namespace 'AmplifyUI' as a type.

angular: 7.1.3
typescript: 3.16
aws-amplify: 1.1.17
aws-amplify-angular: 2.14

### LINKED ISSUE #2365: Failing builds in Typescript project using aws-amplify/api since this line was added
**Describe the bug**
Since the addition of the [linked line](https://github.com/aws-amplify/amplify-js/blob/be5ec549d8c30b42e01701644d9e12aed2b9e39d/packages/api/src/types/index.ts#L13), our builds have been failing with:

```
<truncated>/@aws-amplify/api/lib/types/index.d.ts
ERROR in <truncated>/node_modules/@aws-amplify/api/lib/types/index.d.ts(1,30):
TS7016: Could not find a declaration file for module 'graphql/language/ast'. '<truncated>/node_modules/graphql/language/ast.js' implicitly has an 'any' type.
  Try `npm install @types/graphql` if it exists or add a new declaration (.d.ts) file containing `declare module 'graphql/language/ast';`
```

We are using typescript@3.2.2 and @aws-amplify/api@1.0.23. Our tsconfig.json is: 

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "outDir": "build/dist",
    "module": "esnext",
    "target": "es5",
    "lib": ["es6", "dom"],
    "sourceMap": true,
    "allowJs": true,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "src",
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "suppressImplicitAnyIndexErrors": true,
    "noUnusedLocals": false,
    "typeRoots": [
      "./typings/", "./node_modules/@types/"
    ]
  },
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "src/setupTests.ts"
  ]
}
```

I can work-around the issue with the following steps
1. Add @types/graphql as a dev dependency
2. Change typsecript lib ( tsconfig.json: compilerOptions.lib: ["esnext", "dom"])

At the moment, we aren't using graphql, so the graphql dependency is a little unfortunate.

**To Reproduce**
1. Create a new typescript app (we use [create-react-app-ts](https://github.com/wmonk/create-react-app-typescript))
2. Add a dependency on @aws-amplify/api
3. Try to build

**Expected behavior**
When not using graphql, I shouldn't have to use graphql types.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser: Any
 - Version Any

### PR DESCRIPTION
**Pull request by manueliglesias**
- Add type guard for DocumentNode
- Refactor `getGraphqlOperationType`

Fixes #2365
Fixes #2362

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nihakue** [comment]
Will this cause issues with consumers who aren't including 'esnext' in their tsconfig.json `lib`? See https://github.com/apollographql/graphql-subscriptions/issues/83 `AsyncIterator` might not be used in the version of @types/graphql you're adding, but if it is this change could cause a new issue.

**Comment by samuelcastro** [comment]
@manueliglesias When should we expect the merge? It'll fix https://github.com/aws-amplify/amplify-js/issues/2365 which is blocking me right now.

**Comment by cornr** [comment]
Any news on this PR?

File: packages/api/package.json [hunk 1]
```diff
@@ -53,6 +53,7 @@
     "@aws-amplify/auth": "^1.2.13",
     "@aws-amplify/cache": "^1.0.20",
     "@aws-amplify/core": "^1.0.20",
+    "@types/graphql": "0.13.0",
```
Comment (jordanranz) [inline packages/api/package.json]: Move these types to devDependencies.

**Pull request closed without merging**